### DISCUSSION TITLE
Plots look downsampled compared to the original data

### DISCUSSION THREAD

**Discussion by cyasing** [pybamm-team/PyBaMM#2629] [Q&A] [answered]
I simulated the WLTC drive cycle (using my own data), with a power function input. The current, voltage and power plots look really smoothed out compared to the original data that I have (also plotted for comparison). What could be the possible reasons?

Plots:

1. Original drive cycle:
![image](https://user-images.githubusercontent.com/59028917/215049620-dec54940-5693-442d-b866-c6cb411ab5e0.png)

2. Simulation Result:
![Screenshot 2023-01-27 140612](https://user-images.githubusercontent.com/59028917/215049948-28ada7e6-20cb-40f9-af4a-0317e55c4042.png)

3. Power Function:
![Screenshot 2023-01-27 143830](https://user-images.githubusercontent.com/59028917/215050111-4c72b2c2-5b0d-4e88-ae4a-8c5e3be787cf.png)

You can see that some of the trends in the data are captured, but yeah, not completely accurate.

**Comment by YannickNoelStephanKuhn** [comment]
Power inputs are handled via an algebraic function; that is, an extra equation gets solved for the current that reproduces the requested power input. So you probably just need to increase the temporal resolution of your simulation or lower the absolute or relative tolerances. That way, the current needed to mimic the power input gets computed more accurately.

**Reply by cyasing** [reply]
I see. So for the resolution, should I increase the values in t_eval?

**Reply by YannickNoelStephanKuhn** [reply]
You provide a `pybamm.CasadiSolver(rtol=, atol=, root_tol=, dt_max=)` to the `solver` argument of  `pybamm.Simulation`.

**Reply by cyasing** [reply]
I tried this, in various configurations, and... this is the best I could get if not for the defaults:

![image](https://user-images.githubusercontent.com/59028917/215078585-5558be1f-706e-494d-8443-20dd0d05e944.png)

![image](https://user-images.githubusercontent.com/59028917/215078617-97dfdf9f-7309-4265-93a8-5ade71302b4a.png)

From: solver = pybamm.CasadiSolver(atol=1e-3, rtol=1e-3, mode="safe", dt_max=1e-6)

**Reply by YannickNoelStephanKuhn** [reply]
Those are the defaults, yes. But you forgot `root_tol`, which is important for the algebraic equations. Copy that default value as well (it was 1e-3 as well I think) and then lower atol, rtol and root_tol in steps of one order of magnitude: 1e-4, 1e-5, 1e-6, ..., until the graph starts looking right.

**Reply by cyasing** [reply]
Alright, I will get to it.

**Reply by cyasing** [reply]
> Those are the defaults, yes. But you forgot `root_tol`, which is important for the algebraic equations. Copy that default value as well (it was 1e-3 as well I think) and then lower atol, rtol and root_tol in steps of one order of magnitude: 1e-4, 1e-5, 1e-6, ..., until the graph starts looking right.

So, it looks like there is nothing really that can be improved upon by changing tolerance values. The graph only went back to the form of my original question when I changed solver from slow to fast mode, since slow mode somehow resulted in the weird form. Values below 1e-12 for atol and root_tol raise error, although rtol and dt_max can go much lower. No improvements whatsoever.

**Reply by cyasing** [reply]
> Those are the defaults, yes. But you forgot `root_tol`, which is important for the algebraic equations. Copy that default value as well (it was 1e-3 as well I think) and then lower atol, rtol and root_tol in steps of one order of magnitude: 1e-4, 1e-5, 1e-6, ..., until the graph starts looking right.

It seems to me that the real issue is with the power interpolant, which, even though has been passed with the original data, returns the smoothed out plot for "Power [W]" parameter. Is there anything to try in that regard?

**Reply by cyasing** [reply]
> Power inputs are handled via an algebraic function; that is, an extra equation gets solved for the current that reproduces the requested power input. So you probably just need to increase the temporal resolution of your simulation or lower the absolute or relative tolerances. That way, the current needed to mimic the power input gets computed more accurately.

Can you tell me more about the 'algebraic equation'? I want to understand better how the current is obtained without any other information (as it looks like to me). The class "PowerFunctionControl" seems to just declare P = V * I and that is all it takes to solve? A reference for this is fine, too.

**Reply by YannickNoelStephanKuhn** [reply]
"All it takes" is the impression you get because the hard work is hidden from view here: the DAE (Differential-Algebraic Equation) solver in CasADi is good enough to find the solution satisfying both the differential equations and P = V * I.
https://web.casadi.org/docs/#document-daebuilder

**Reply by cyasing** [reply]
Sorry about that, I meant "all it takes" in perspective of setting up the power to current conversion using CasADi. I was aware of CasADi solver from our earlier discussion, just wasn't sure how P = V * I was enough information to get current values. So these DAE's are the SPM/DFN/etc. models that we choose right?

**Reply by YannickNoelStephanKuhn** [reply]
The DAEs are the SPM/SPMe/DFN models, yes. P = V * I is enough information because CasADi solves for the current that satisfies P_applied - V * I = 0. In terms of internal equation structure, V is a function of current. So it's not that different from solving for the current that satisfies V_applied - V = 0.

**Reply by cyasing** [reply]
Nice, that makes sense. Thank you, it's been bugging me since forever!

**Reply by valentinsulzer** [reply]
We never wrote a paper on this, but luckily @MarcBerliner did https://iopscience.iop.org/article/10.1149/1945-7111/ac9a80/pdf (he calls it "General Operating Modes")

**Reply by cyasing** [reply]
@tinosulzer I see. That's neat!

**Comment by valentinsulzer** [comment, marked as answer]
Looks like a time scaling issue. How are you defining the interpolant? It should be something like this, but for power:
```python3
# create interpolant - must be a function of *dimensional* time
timescale = param.evaluate(model.timescale)
current_interpolant = pybamm.Interpolant(drive_cycle[:, 0], drive_cycle[:, 1], timescale * pybamm.t)
```
You won't need to specify the timescale any more once we remove the non-dimensionalization. But for now you do.

**Reply by cyasing** [reply]
It's the same syntax:
```
drive_cycle_power = pd.read_excel("wltc_power.xlsx",skiprows=0).to_numpy()

# Create interpolant for drive cycle
timescale = param.evaluate(model.timescale)
power_interpolant = pybamm.Interpolant(drive_cycle_power[:, 0], drive_cycle_power[:, 1], timescale * pybamm.t)
param.update({"Power function [W]": power_interpolant}, check_already_exists=False) # Drive Cycle

sim = pybamm.Simulation(model, parameter_values=param, solver=solver, geometry=geometry)
#t_eval = drive_cycle_power[:,0]
t_eval = np.linspace(0,1800,1801)
solution = sim.solve(t_eval)
```

**Reply by valentinsulzer** [reply]
Strange. It really looks like it's only using the first 50 seconds of the drive cycle, but stretched to 1800. What's the value of `timescale`?

Can you plot `power_interpolant.evaluate(t_eval)`?

**Reply by cyasing** [reply]
Oh no, here's what I got:

![image](https://user-images.githubusercontent.com/59028917/215228670-1ef90073-b17b-47f3-ad1e-98b8455f5cd6.png)

**Reply by cyasing** [reply]
Apparently, all I needed to do was to eliminate the extra columns of data in my drive cycle file! I also had a bit of calculation on that data in the script, I moved it into excel. Then, it simulated well! That's so strange to me.

![image](https://user-images.githubusercontent.com/59028917/215234571-153ca5fd-1f95-470b-92d4-de505308bbcb.png)

The 'power_interpolant.evaluate(t_eval)' did not get better, though.

**Reply by cyasing** [reply]
> power_interpolant.evaluate(t_eval)

Is my result something to be worried about? It does look like it to me.

**Reply by valentinsulzer** [reply]
Do `power_interpolant.evaluate(t_eval/timescale)`. The power interpolant is stored as a function of dimensionless time

**Reply by cyasing** [reply]
That makes sense. Here, I plotted it:
![image](https://user-images.githubusercontent.com/59028917/215690801-15850dd4-fc18-4aa7-ad4a-b6a2ea284ada.png)
Looks like it's fine now. Thanks for all your help!
